src/Microsoft.AspNet.Hosting/Internal/WebApplication.cs

```diff
@@ -98,14 +127,14 @@ public virtual IApplication Start()
             _applicationLifetime.NotifyStarted();
             logger.Started();
 
-            return new Application(ApplicationServices, Server.Features, new Disposable(() =>
-            {
-                logger.Shutdown();
-                _applicationLifetime.StopApplication();
-                Server.Dispose();
-                _applicationLifetime.NotifyStopped();
-                (_applicationServices as IDisposable)?.Dispose();
-            }));
+            return new Disposable(() =>
```

Comment:
For addresses we should expose those as a first-class controllable thing, like environment. Also, when you call .Build() there's waaay too much of the processing being done. Creating the server factory, and getting the features, and populating the application builder should all happen within the Start method.

So that would look like
```
            new WebApplicationBuilder()
                .UseConfiguration(commandLineArgs: args)
                .UseEnvironment("Production")
                .UseAddresses("http://localhost:5000", "https://localhost:5001")
                .Run();
```

Heck, at that point we should probably add extension methods overloads for Start and Run which take addresses. (Kind of like node's final method call of .listen takes the port).

```cs
    public class StartupExternallyControlled
    {
        private readonly List<string> _addresses = new List<string>();
        private IWebApplication _application;

        public StartupExternallyControlled()
        {
        }

        public void Start()
        {
            _application = new WebApplicationBuilder()
                .Start(_addresses.ToArray());
        }

        public void Stop()
        {
            _application.Dispose();
        }

        public void AddUrl(string url)
        {
            _addresses.Add(url);
        }
    }
```

```cs
    public class StartupConfigureAddresses
    {
        // Entry point for the application.
        public static void Main(string[] args)
        {
            new WebApplicationBuilder()
                .UseConfiguration(
                    hostingJsonFile: "hosting.json", 
                    commandLineArgs: args)
                .UseEnvironment("Production")
                .Run("http://localhost:5000", "https://localhost:5001");
        }
    }
```